BOARD, SOC, DTS, ARCH ROOT in Zephyr modules.

tejlmand opened the pull request:
This PR contains a series of commits that:
- Cleanup usage of SOC dir.
- Add support for a list of SOC_ROOTs, instead of only a single root
- Kconfig handling of multiple SOC_ROOTs
- Add support for a list of ARCH_ROOTs, instead of only a single root
- Minor fixes where code were wrongly using `${ZEPHYR_BASE}/arch` which doesn't work with out-of-tree archs, instead the correct `${ARCH_DIR}` is now used.

And then finally, support for specifying BOARD, DTS, ARCH, and SOC root in Zephyr modules, `zephyr/module.yml` files.

This fixes: #25215

tejlmand committed:
xtensa: SOC_DIR usage cleanup.

This commit is a cleanup of the SOC_DIR usage.
This cleanup is a preparation for supporting a list of SOC_ROOT instead
of just allowing one SOC_ROOT.

Supporting a list of SOC_ROOTs allows for placing of SOC in Zephyr
modules. It also aligns how BOARD_ROOT supports a list, and thus usage
of n_ROOT in Zephyr becomes more consistent.

This commit introduces the following changes:
- soc/xtensa/intel_apl_adsp/bootloader.cmake removed.
  This file is not included elsewhere in the build system, and appears
  to be leftover from #25133. Almost identical content is found in
  `soc/xtensa/intel_apl_adsp/commonbootloader.cmake`
- Changed xtensa/intel_apl_adsp to named library. Using a named library
  allow fetching library files based on library name without the need to
  know build path (and thus removes the need for knowing `${SOC_DIR}`).
- Changed SOC_DIR/ARCH/SOC_FAMILY to use CMAKE_CURRENT_LIST_DIR for
  configure time commands, as CMake code is already located inside this
  path.
- Using generator expression for library files from other CMake targets.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: kconfig: support for multiple SOC_ROOT

This commit introduces support for multiple SOC_ROOT.

This means that additional SOC_ROOTs specified using -DSOC_ROOT as
argument to CMake will be forming a list together with ${ZEPHYR_BASE}.

This allows for greater flexibility, as developers can now specify
multiple out-of-tree SoCs and not worry about the SoC used for the
board they compile for.

Also it avoid code, such as:
if(BOARD STREQUAL my_board_using_out_of_tree_soc)
  set(SOC_ROOT some/out/of/tree/soc/path)
endif()
in application CMakeLists.txt.

Finally, allowing multiple SOC_ROOTs prepares for specifying SOC_ROOTs
in Zephyr modules.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: relocation of generated Kconfig file for Zephyr modules

This commit relocates the generated Kconfig.modules file from
${CMAKE_BINARY_DIR} into ${CMAKE_BINARY_DIR}/Kconfig.

This is done as preparation of multiple SOC and ARCH root support.

A single Kconfig folder inside ${CMAKE_BINARY_DIR} helps to keep the
build folder tidy.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: fix include directories to work with out-of-tree arch

Include directories for ${ARCH} is not specified correctly.
Several places in Zephyr, the include directories are specified as:
${ZEPHYR_BASE}/arch/${ARCH}/include
the correct line is:
${ARCH_DIR}/${ARCH}/include
to correctly support out of tree archs.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: allowing list of ARCH_ROOTs

This commit aligns ARCH_ROOT to conform to BOARD_ROOT, DTS_ROOT,
SOC_ROOT that all allows list of directories.

Also it avoid code, such as:
if(BOARD STREQUAL my_board_using_out_of_tree_arch)
  set(ARCH_ROOT some/out/of/tree/arch/path)
endif()
in application CMakeLists.txt.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
modules: adding support for ROOTs definitions in zephyr/module.yml

Fixes: #25215

This commit introduces the possibility of defining ROOTs in a Zephyr
module and have it automatically appended to list of other ROOTs.
Supported with this commit:
- BOARD_ROOT
- SOC_ROOT
- DTS_ROOT
- ARCH_ROOT

In order to support this in Zephyr module files, the detection of west
has been moved to dedicated west.cmake file and included immediately
after python.cmake.

Also the inclusion of zephyr_modules.cmake has moved before first use
of BOARD_ROOT.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nashif commented:
nice, we do need a test for all of this... some dummy module or whatever that can verify all of this works as expected.

tejlmand commented:
> nice, we do need a test for all of this... some dummy module or whatever that can verify all of this works as expected.

Agreed, currently I have adopted this into this principle https://github.com/nrfconnect/sdk-nrf/pull/2566 for early testing.
so that works as additional test.

But I want to add a proper test in Zephyr itself.
Just wanted to give people a chance to give feedback on the rest.

mbolivar-nordic commented on cmake/kconfig.cmake:
```diff
@@ -5,6 +5,22 @@
 file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/kconfig/include/config)
 
+# Support multiple SOC_ROOT
+set(WRITE WRITE)
```
Kind of tricky when reading this for the first time... why not call this OPERATION or WRITE_OP or something?

tejlmand replied:
`OPERATION` is good, will use.

mbolivar-nordic replied:
> `OPERATION` is good, will use.

Still planning on doing this?

tejlmand replied:
done

mbolivar-nordic commented on scripts/ci/check_compliance.py:
```diff
@@ -256,6 +256,28 @@ def get_modules(self, modules_file):
+
+        """
+
+        soc_defconfig_file = os.path.join(tempfile.gettempdir(), "Kconfig.soc.defconfig")
```
Are you sure about this? There's only one unique file for this in the entire CI build, and it's /tmp/Kconfig.soc.defconfig? I don't know this code, but this seems like a concurrency problem waiting to happen. What am I missing here?

tejlmand replied:
That could be, i'm not to familiar with the new compliance test, to now if that is indeed a risk.
But this code simply follows the existing pattern in the compliance test:
https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/ci/check_compliance.py#L296
```
        # For multi repo support
        self.get_modules(os.path.join(tempfile.gettempdir(), "Kconfig.modules"))
```

However, following an already used pattern elsewhere in this file, is no guarantee that it is correct.
@nashif Could there be a risk of concurrency problem just waiting to happen ?

mbolivar-nordic replied:
>  @nashif Could there be a risk of concurrency problem just waiting to happen ?

I guess we'll find it the hard way if so. Since this PR is just keeping things consistent with how this file works, I'm assuming it's OK.

mbolivar-nordic commented on cmake/kconfig.cmake:
```diff
@@ -5,6 +5,22 @@
+set(WRITE WRITE)
+foreach(root ${SOC_ROOT})
+  file(${WRITE} ${KCONFIG_BINARY_DIR}/Kconfig.soc.defconfig
+       "osource \"${root}/soc/$(ARCH)/*/Kconfig.defconfig\"\n"
```
Can you provide some rationale for using 'osource' instead of checking if the file exists at configuration time and using 'source' if it does? I could see benefits to either approach and I'm curious why you chose this one.

I'm slightly concerned that files which did not exist at build system generation time can start to affect kconfig's behavior after generation, as 'osource' checks for existence when the user runs menuconfig, I think. Any issues there?

tejlmand replied:
This existing Kconfig code looks as:
```
source "$(SOC_DIR)/$(ARCH)/*/Kconfig.defconfig"
```
means that there can only be a single SOC root, which prevents support of multiple SOC roots.
Now, the `ARCH` is deduced from the board folder so that is known, but it is not enough, cause we don't know if we need to use a soc in SOC_ROOT 1, 2, 3, or `<n>`.
So we need all SOC roots in Kconfig, to be able to determine the right SOC.

Reason for using `osource`:
Example 1, someone adds a new `arm` SOC, called `mysoc`, and adds `-DSOC_ROOT=<path-to-additional-soc-root>` .
Now, as `arm` is a known arc in both `${ZEPHYR_BASE}/soc/arm` and in `<path-to-additional-soc-root>/soc/arm` then source would work.
Example 2,  User also creates `myarch`, then `${ZEPHYR_BASE}/soc/myarch` will fail if using `source` cause that folder doesn't exist, but `<path-to-additional-soc-root>/soc/myarch` works with source.
Example 2b, The opposite of 2, adding additional `arm` soc, but user decides to build for an `xtensa` soc.

But using `osource` then all cases above passes, for example `<path-to-additional-soc-root>/soc/${ARCH}`, even though `xtensa` is not present (it's present in `${ZEPHYR_BASE}/soc/xtensa`).
This means we allow users to always add additional SOC_ROOTs, even when building for an arch not present at that SOC root.
This is especially useful when allowing Zephyr modules with SOC root.

We could potentially check for existence of just any random `Kconfig.defconfig` file under `<root>/soc/${ARCH}/*` and not add the line if no file at all if found.
But that would move additional Kconfig logic into CMake, and not have it in Kconfig.
I find this knowledge to be Kconfig specific knowledge, and thus prefer it in Kconfig.
(I would have preferred an ability in Kconfig to loop a list of folders, but that is not possible, hence the reason to generate additional Kconfig files in CMake)

Also, using `osource` keeps the behavior consistent with today.
If you run:
```
cmake ... -Bbuild .
# build/zephyr/.config has been generated
mkdir ${ZEPHYR_BASE}/soc/arm/my_soc
echo 'config SOME_SETTING\n    default "myval"' > ${ZEPHYR_BASE}/soc/arm/my_soc/Kconfig.defconfig
```
then that file was not used for the first `.config` generation, but will be taken into consideration on later Kconfig invocation (including menuconfig), due the the `*` globbing:
```
source "$(SOC_DIR)/$(ARCH)/*/Kconfig.defconfig"
```

Using `osource` ensures the same behaviour, if user does:
```
cmake ... -Bbuild .
# build/zephyr/.config has been generated
# ${other-soc-root}/soc/arm does not exists, let's create it
mkdir ${other-soc-root}/soc/arm
# Now create my_soc folder
mkdir ${other-soc-root}/soc/arm/my_soc
echo 'config SOME_SETTING\n    default "myval"' > ${ZEPHYR_BASE}/soc/arm/my_soc/Kconfig.defconfig
```

If we had not added `osource \"${other-soc-root}/soc/$(ARCH)/*/Kconfig.defconfig\"\n` to the Kconfig in first place, then the behavior would be different depending on whether just a single `Kconfig.defconfig`  existed somewhere or not.

mbolivar-nordic replied:
> Also, using osource keeps the behavior consistent with today.

This more than anything else is a good reason to keep it this way IMO. Thanks for explaining.

mbolivar-nordic commented on scripts/ci/check_compliance.py:
```diff
@@ -284,7 +306,7 @@ def parse_kconfig(self):
         os.environ["BOARD_DIR"] = "boards/*/*"
         os.environ["ARCH"] = "*"
-        os.environ["CMAKE_BINARY_DIR"] = tempfile.gettempdir()
+        os.environ["KCONFIG_BINARY_DIR"] = tempfile.gettempdir()
```
I guess this is using /tmp too, so maybe it's fine...

tejlmand replied:
yeah, it is existing behavior, but doesn't mean it is safe.
See other comment.

mbolivar-nordic commented on Kconfig.zephyr:
```diff
@@ -22,10 +22,10 @@ endmenu
+source "$(KCONFIG_BINARY_DIR)/Kconfig.soc.defconfig"
 
 source "boards/Kconfig"
-source "$(SOC_DIR)/Kconfig"
```
I see SOC_DIR still mentioned in doc/application/index.rst; is that OK?

Are you planning to update documentation later if this approach is accepted? If so, maybe consider adding DNM or marking as a draft.

I still see SOC_DIR getting set in boilerplate.cmake, so I'm not sure why it's not available for use here. I see that it's no longer set in the environment in kconfig.cmake, but I don't understand why if we still know where it is.

tejlmand replied:
yes, the docs still need to be updated, and thanks for the heads-up on SOC_DIR in the docs.
Will keep an extra eye on that.

I'm simply waiting for first feedback before updating the docs, especially on the yaml changes, as those are most important in that regard.

> I still see SOC_DIR getting set in boilerplate.cmake, so I'm not sure why it's not available for use here. I see that it's no longer set in the environment in kconfig.cmake, but I don't understand why if we still know where it is.

Because Kconfig determines the soc we are using, and once we know the soc, then we can set the soc dir.
But that means we need all soc roots available for Kconfig.

But for safety, the `${SOC_DIR}` should only be needed for `add_subdirectory` in CMake.
(it's used a limited number of places for includes and linker script, but hopefully that can be cleaned up later)

mbolivar-nordic commented on doc/CMakeLists.txt:
```diff
@@ -195,6 +197,17 @@ else()
+file(WRITE ${KCONFIG_BINARY_DIR}/Kconfig.soc.defconfig
+     "osource \"${ZEPHYR_BASE}/soc/$(ARCH)/*/Kconfig.defconfig\"\n"
+)
+file(WRITE ${KCONFIG_BINARY_DIR}/Kconfig.soc
```
Maybe call this one Kconfig.soc_selection, and `s/Kconfig.soc.arch/Kconfig.soc_config/` also?

I'm not finding these file names intuitive.

tejlmand replied:
Not sure what `selection` refers to, as we have not yet selected the soc that will be used.

The reason for that names, originates from existing Kconfig names:
https://github.com/zephyrproject-rtos/zephyr/blob/master/Kconfig.zephyr#L25
https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/Kconfig#L6
https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/Kconfig#L11
```
source $(SOC_DIR)/$(ARCH)/*/Kconfig.defconfig  --> Kconfig.soc.defconfig  (defconfig a bit too generic in build folder, so added soc)
source "$(SOC_DIR)/$(ARCH)/*/Kconfig.soc"         --> Kconfig.soc
osource "$(SOC_DIR)/$(ARCH)/Kconfig"                --> Kconfig.soc.arch   (Kconfig.soc already used, and this is soc/arch specific file)
```
We could rename `Kconfig.soc` into `Kconfig.soc.soc`.

> `s/Kconfig.soc.arch/Kconfig.soc_config/`

not sure that name is better. All Kconfig file contains config settings.

Overview of the files and their content (`${SOC_DIR}` removed):
- `$(ARCH)/*/Kconfig.soc`: SOC specific configuration, as example `arm/nordic_nrf/Konfig.soc` loads `arm/nordic_nrf/Konfig.series` which loads nRF52 series specific settings. So this is specific to the soc, not the arch.
  Hence Kconfig.soc is good name in build folder when refering to this file, imho
- `$(ARCH)/Kconfig`: load architecture specific configuration for all SOC in this arch, as example `arm/Kconfig` loads specific configuration settings valid for arm socs, but not for xtensa socs. So specific for all soc belonging to this arch.
    Hence `Kconfig.soc.arch` is good name in build folder when refering to this file, imho
- `$(ARCH)/*/Kconfig.defconfig`: default settings for this soc.
    Hence Kconfig.soc.defconfig is good name when refering to this file, imho

mbolivar-nordic replied:
Thanks for the justification of the names. No objections.

mbolivar-nordic commented on cmake/app/boilerplate.cmake:
```diff
@@ -115,9 +115,19 @@ include(${ZEPHYR_BASE}/cmake/version.cmake)  # depends on hex.cmake
+# Also create the Kconfig binary dir for generated Kconf files.
+#
+set(KCONFIG_BINARY_DIR ${CMAKE_BINARY_DIR}/Kconfig)
+file(MAKE_DIRECTORY ${KCONFIG_BINARY_DIR})
```
Doesn't this still need to happen after the `if(${CMAKE_CURRENT_SOURCE_DIR} STREQUAL ${CMAKE_CURRENT_BINARY_DIR})` check below? Otherwise it looks like it would dirty the source tree.

tejlmand replied:
good catch, thanks.
In-tree builds is evil, but should be handled correctly.

mbolivar-nordic commented on scripts/zephyr_module.py:
```diff
@@ -177,6 +207,9 @@ def main():
     parser.add_argument('--cmake-out',
                         help="""File to write with resulting <name>:<path>
                              values to use for including in CMake""")
+    parser.add_argument('--settings-out',
```
This looks identical in purpose to --cmake-out; do we really need a separate file? If so, why?

tejlmand replied:
Yes, we should keep those separated.

Although both files are currently used only by CMake, then the cmake out file contains module name and path for inclusion into the build tree itself, with `add_subdirectory` (They contain the path to the CMakeLists.txt file).
And therefore is very CMake specific.

The settings file is simply a file containing build settings from each Zephyr module.
This could be useful for other things in future, for example documentation generation, if we want to create a page with boards located at other Zephyr modules.

But of course, CMake will use the settings for building 😉 

And then finally, settings file is a `key:value` pair in all lines, but in `--cmake-out` file then all lines in  are values (The keys are `module_name:path`).

mbolivar-nordic commented on scripts/zephyr_module.py:
```diff
@@ -50,6 +50,22 @@
+        mapping:
+          board_root:
+            required: false
+            type: str
```
Nice, I like this approach.

Looks like these are relative paths from the root of the module which are made absolute below in process_settings().

Maybe it would be a good time to discuss https://github.com/zephyrproject-rtos/zephyr/issues/23825 in this context. What should a relative path mean when set by the user? Relative to the application directory, still? Will the various ROOT support sections be able to handle that with this change? Not in this PR, just in general.

tejlmand replied:
Agree, #23825 is worth a good discussion, but let's try not to go to deep with that in this PR.
As a Zephyr module can be located anywhere, and relocated at will, then the only thing we know inside the module is relative to its own root.

But for a user providing a `-DBOARD_ROOT=<relative-path>` then I would say that should be relative to his current working dir.
So if he is in:
```
cd /tmp
cmake ... -Dx_ROOT=<relative-path> -Bbuild /path/to/app
```
then path is relative to `/tmp`, and if he does:
```
cd /path/to/app
cmake ... -Dx_ROOT=<relative-path> -Bbuild .
```
then relative to `/path/to/app`

> Will the various ROOT support sections be able to handle that with this change? Not in this PR, just in general.

In this PR, and in relation to Zephyr modules, it will work.
In relation to `-Dx_ROOT=<relative-path>`, then I would need to dig deeper into how that is currently handled.

mbolivar-nordic replied:
>  As a Zephyr module can be located anywhere, and relocated at will, then the only thing we know inside the module is relative to its own root.

Yep, I agree that is the right approach in this particular case. I just wanted to think about whether we're taking any steps in this PR that would preclude solving #23825.

I see now that we can do an `if(IS_ABSOLUTE ...)` check later on, so I think the answer to that question is "no". We can treat relative paths in these lists as undefined behavior for now, and define proper behavior for them later.

> Agree, #23825 is worth a good discussion, but let's try not to go to deep with that in this PR.

What do you think about at least issuing a CMake warning `if(NOT IS_ABSOLUTE ${some_root})` in the loops that go over these ROOT variables?

Since we are making the behavior of these variables more complex here and the user-visible portions are starting to use relative paths, it might be nice to give the user some hints if they are doing something "by hand" that we know won't work.

tejlmand replied:
> > As a Zephyr module can be located anywhere, and relocated at will, then the only thing we know inside the module is relative to its own root.
> 
> Yep, I agree that is the right approach in this particular case. I just wanted to think about whether we're taking any steps in this PR that would preclude solving #23825.

No, we're not.
The `zephyr_module.py` will always provide full path, and using `set(DTS_ROOT <path>)` or using `-DDTS_ROOT=<path>` does not impact that.

#23825 is an independent issue.

> 
> I see now that we can do an `if(IS_ABSOLUTE ...)` check later on, so I think the answer to that question is "no". We can treat relative paths in these lists as undefined behavior for now, and define proper behavior for them later.
> 
> > Agree, #23825 is worth a good discussion, but let's try not to go to deep with that in this PR.
> 
> What do you think about at least issuing a CMake warning `if(NOT IS_ABSOLUTE ${some_root})` in the loops that go over these ROOT variables?
> 
> Since we are making the behavior of these variables more complex here and the user-visible portions are starting to use relative paths, it might be nice to give the user some hints if they are doing something "by hand" that we know won't work.

I think there is a big difference in using `-Dx_ROOT=<relative-path>` and `set(DTS_ROOT <relative-path>)` compared to a Zephyr module.
The Zephyr modules are always (and has always) been using relative path from their module root.
This goes for the new build settings, but also for the existing CMake and Kconfig settings.
Then `zephyr_module.py` ensures to expand those settings into absolute path for CMake and Kconfig.

Whereas #23825 only covers relative path in relation to CMake.

mbolivar-nordic reviewed (commented):
I like the final changes to the module YAML file schema. Restricting this to specific settings seems like the right thing to do. I'm still taking this in and have various questions that I'll just leave here since we've got long round trips.

Thanks!

daor-oti reviewed (approved):
Nice cleanup and additions.

tejlmand commented:
Note: `TOOLCHAIN_ROOT` will not be reworked in this PR, cause toolchain abstraction work is ongoing here: https://github.com/zephyrproject-rtos/zephyr/pull/24851

Until that work is completed, the `TOOLCHAIN_ROOT` will keep its current behavour.

carlescufi commented:
@tejlmand this needs rebasing, and you probably need to include changes from:
https://github.com/zephyrproject-rtos/zephyr/commit/ef3c5e551676278ccc1f42ecb572838144c5e33a#diff-0caca5b5c60e0efd6c4221fa6d1f7b61
and
https://github.com/zephyrproject-rtos/zephyr/commit/a043d48c54726082521f3e335d807dc649cc02f3#diff-0caca5b5c60e0efd6c4221fa6d1f7b61

mbolivar-nordic reviewed (approved):
I have no remaining blocking review comments.

tejlmand commented:
> > nice, we do need a test for all of this... some dummy module or whatever that can verify all of this works as expected.
> 
> Agreed, currently I have adopted this into this principle [nrfconnect/sdk-nrf#2566](https://github.com/nrfconnect/sdk-nrf/pull/2566) for early testing.
> so that works as additional test.
> 
> But I want to add a proper test in Zephyr itself.
> Just wanted to give people a chance to give feedback on the rest.

I'm working on tests for Zephyr modules, especially related to root features in this PR.
However this requires a bit more work in sanitycheck / sanitylib due to how board roots are handled there, so I will keep that work outside of this PR, to allow people to start using this feature.

Will post dedicated PR later with testing of Zephyr modules.

tejlmand committed:
doc: zephyr module description restructure

This commit restructures the Zephyr module description.
It places the description of the module yaml file in a dedicated section
as this file is common to module inclusion both when using `west` and
without `west`.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
doc: Zephyr module build settings

This commit adds description of the build setting support in the
module.yml file.

It is described how additional board_root, soc_root, dts_root, and
arch_root can be appended to the build system.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
docs updated with description of the root feature.
